llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp

```diff
@@ -106,6 +107,46 @@ static bool replaceWithLLVMIR(FPBuiltinIntrinsic &BuiltinCall) {
   return true;
 }
 
+// This function lowers llvm.fpbuiltin. intrinsic functions with 3.0 max-error
+// attribute to the appropriate nvvm approximate intrinsics if it's possible.
+// If it's not possible - fallback to instruction or standard C/C++ library LLVM
+// intrinsic.
+static bool
+replaceWithApproxNVPTXCallsOrFallback(FPBuiltinIntrinsic &BuiltinCall) {
+  IRBuilder<> IRBuilder(&BuiltinCall);
+  SmallVector<Value *> Args(BuiltinCall.args());
+  Value *Replacement = nullptr;
+  auto *Type = BuiltinCall.getType();
+  // For now only add lowering for fdiv and sqrt. Yet nvvm intrinsics have
+  // approximate variants for sin, cos, exp2 and log2.
+  // For vector fpbuiltins for NVPTX target we don't have nvvm intrinsics,
+  // fallback to instruction or standard C/C++ library LLVM intrinsic. Also
+  // nvvm fdiv and sqrt intrisics support only float type, so fallback in this
+  // case as well.
+  switch (BuiltinCall.getIntrinsicID()) {
+  case Intrinsic::fpbuiltin_fdiv:
+    if (Type->isVectorTy() || !Type->getScalarType()->isFloatTy())
```

Comment:
This is a bit of a nit-picking, so feel free to ignore. Since both cases are identical, bar the intrinsic ID, I'd be tempted to only set the ID in the body of `case`, and do the type check/create intrinsic afterwards.

Reply:
I see your point and I agree that it would be better - I was thinking to do it like you have said initially, but then I've intentionally left it inside the switch case. The reason is (see the comment above: "For now only add lowering for fdiv and sqrt."): there are other nvvm intrinsics, that might or might not support vectors and/or doubles. For example looking at NVPTXIntrinsics.td I see, that there are both `lg2.approx.f32` and `lg2.approx.f64` overloads of `lg2` intrinsic. In this patch I'm solving a narrow issue, but I can imagine this switch case being expanded in the future, so for now those ifs are moved inside.